testlib/src/main/java/com/diffplug/spotless/StepHarness.java

```diff
@@ -23,16 +23,22 @@
 import org.assertj.core.api.Assertions;
 import org.junit.Assert;
 
-/** An api for adding test cases. */
+/**
+ * An api for adding test cases.
+ */
 public class StepHarness implements AutoCloseable {
```

Comment:
I think we need to revert these changes to StepHarness.  You are not doing it in this PR, but we've had quite a few PR's over the years (including a recent one) where people suck the content from the file, rather than from the String.  That breaks the Spotless model of composing multiple formatters, autofixing idempotence problems, etc.  Inevitably, the way we catch this is that they have made the same changes you have made here.  Probably I should add a comment saying "don't change this!!!".

If you need a test where the filenames have been set, feel free to add a class `StepHarnessWithFile`.  But the `StepHarness` does not have files on purpose.

Reply:
Alright, I also wasn't sure when changing the API if that is the way to go, but it felt more cumbersome to have two separate classes doing (almost) exactly the same thing...

Reply:
You're right that it is a bit cumbersome.  My reasoning is that it's good to keep the simple things simple.  Spotless gets a lot of benefit out of the `Function<String, String>` abstraction, and one of those is that you can test a formatter and not worry about where the strings are coming from.

We have the `File` escape hatch because there are good reasonable cases, such as `prettier`, where you want to provide the underlying `File`.  But once you use that hatch, you now need to worry about where the strings are coming from, and I'd prefer to impose that complexity only on steps which have opted-in to the escape hatch.